chain/actors/actor_storagepower.go

```diff
@@ -273,13 +281,111 @@ func (spa StoragePowerActor) UpdateStorage(act *types.Actor, vmctx types.VMConte
 	if err != nil {
 		return nil, err
 	}
-
 	if !has {
 		return nil, aerrors.New(1, "update storage must only be called by a miner actor")
 	}
 
 	self.TotalStorage = types.BigAdd(self.TotalStorage, params.Delta)
 
+	previousBucket := params.PreviousProvingPeriodEnd % build.ProvingPeriodDuration
+	nextBucket := params.NextProvingPeriodEnd % build.ProvingPeriodDuration
+
+	if previousBucket == nextBucket && params.PreviousProvingPeriodEnd != 0 {
+		nroot, err := vmctx.Storage().Put(&self)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := vmctx.Storage().Commit(old, nroot); err != nil {
+			return nil, err
+		}
+
+		return nil, nil // Nothing to do
+	}
+
+	buckets, eerr := amt.LoadAMT(types.WrapStorage(vmctx.Storage()), self.ProvingBuckets)
+	if eerr != nil {
+		return nil, aerrors.HandleExternalError(eerr, "loading proving buckets amt")
+	}
+
+	if params.PreviousProvingPeriodEnd != 0 { // delete from previous bucket
+		var bucket cid.Cid
+		err := buckets.Get(previousBucket, &bucket)
+		switch err.(type) {
+		case *amt.ErrNotFound:
+			return nil, aerrors.HandleExternalError(err, "proving bucket missing")
+		case nil: // noop
+		default:
+			return nil, aerrors.HandleExternalError(err, "getting proving bucket")
+		}
+
+		bhamt, err := hamt.LoadNode(vmctx.Context(), vmctx.Ipld(), bucket)
+		if err != nil {
+			return nil, aerrors.HandleExternalError(err, "failed to load proving bucket")
+		}
+		err = bhamt.Delete(vmctx.Context(), string(vmctx.Message().From.Bytes()))
+		if err != nil {
+			return nil, aerrors.HandleExternalError(err, "deleting miner from proving bucket")
+		}
+
+		err = bhamt.Flush(vmctx.Context())
+		if err != nil {
+			return nil, aerrors.HandleExternalError(err, "flushing previous proving bucket")
+		}
+
+		bucket, err = vmctx.Ipld().Put(vmctx.Context(), bhamt)
+		if err != nil {
+			return nil, aerrors.HandleExternalError(err, "putting previous proving bucket hamt")
+		}
+
+		err = buckets.Set(previousBucket, bucket)
+		if err != nil {
+			return nil, aerrors.HandleExternalError(err, "setting previous proving bucket cid in amt")
+		}
+	}
+
+	{ // set in next bucket
```

Comment:
I'd pull the delete and set bucket bits out as their own methods. Makes this cleaner to read